subsys/power/device.c

```diff
@@ -110,51 +123,59 @@ int sys_pm_force_suspend_devices(void)
 
 void sys_pm_resume_devices(void)
 {
-	int i;
+	device_idx_t pmi = num_pm - num_susp;
 
-	for (i = 0; i < device_count; i++) {
-		if (!device_retval[i]) {
-			int idx = device_ordered_list[i];
+	num_susp = 0;
+	while (pmi < num_pm) {
+		device_idx_t idx = pm_devices[pmi];
 
-			device_set_power_state(&pm_device_list[idx],
-					DEVICE_PM_ACTIVE_STATE, NULL, NULL);
-		}
+		device_set_power_state(&all_devices[idx],
+				       DEVICE_PM_ACTIVE_STATE,
+				       NULL, NULL);
+		++pmi;
 	}
 }
 
 void sys_pm_create_device_list(void)
 {
 	int count;
-	int i, j;
-	bool is_core_dev;
+	device_idx_t pmi;
 
 	/*
 	 * Create an ordered list of devices that will be suspended.
 	 * Ordering should be done based on dependencies. Devices
 	 * in the beginning of the list will be resumed first.
 	 */
-	device_list_get(&pm_device_list, &count);
+	device_list_get(&all_devices, &count);
+
+	__ASSERT_NO_MSG(count <= (device_idx_t)(-1));
```

Comment:
It would be clearer if UINT16_MAX is used instead of -1.

Reply:
But it would be wrong if `device_idx_t` changed to a larger type.  I could add `#define DEVICE_IDX_MAX UINT16_t` if that would be more acceptable.

Reply:
Sure, it'd be fine. It just looks a bit confusing currently - as if we want `count` to be negative.

Reply:
Added a define right next to the typedef so it should be clear what this idiom does.  Also added a lower bound check in the assert since `device_list_get` forces a signed `count`.